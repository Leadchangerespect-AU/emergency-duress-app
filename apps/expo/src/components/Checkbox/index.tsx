import React, { useId } from "react";
import { Platform, Text, TextProps, View } from "react-native";
import {
  Checkbox as RNCheckbox,
  CheckboxProps as RNCheckboxProps,
} from "expo-checkbox";
import { tv, VariantProps } from "tailwind-variants";

const checkboxStyle = tv({
  slots: {
    checkbox: "mr-2",
    text: null,
  },
  variants: {
    color: {
      primary: {
        checkbox: "accent-primary",
        text: null,
      },
      secondary: {
        pressable: "accent-secondary",
        text: null,
      },
    },
    size: {
      md: {
        checkbox: "px-2.5 py-2.5",
        text: "text-sm",
      },
      lg: {
        checkbox: "px-4 py-4",
        text: "text-md",
      },
    },
    disabled: {
      true: {
        checkbox: "accent-gray-500",
        text: "text-gray-400",
      },
    },
  },
  defaultVariants: {
    color: "primary",
    size: "md",
  },
});

export interface CheckboxProps
  extends Omit<RNCheckboxProps, "disabled" | "children" | "color">,
    Pick<TextProps, "children">,
    VariantProps<typeof checkboxStyle> {
  containerClass?: string;
}

const Checkbox = ({
  disabled,
  color,
  size,
  children,
  containerClass,
  ...props
}: CheckboxProps) => {
  const autoId = useId();
  const { checkbox, text } = checkboxStyle({ disabled, color, size });
  return (
    <View className={`flex-row ${containerClass}`}>
      <RNCheckbox
        className={checkbox()}
        disabled={disabled}
        aria-disabled={disabled}
        color={color === "secondary" ? "#04c8c7" : "#003951"}
        id={autoId}
        role="checkbox"
        aria-checked={props.value}
        {...props}
      />
      {Platform.OS === "web" ? (
        <label className={text()} htmlFor={autoId}>
          {children}
        </label>
      ) : (
        <Text className={text()}>{children}</Text>
      )}
    </View>
  );
};

export default Checkbox;
