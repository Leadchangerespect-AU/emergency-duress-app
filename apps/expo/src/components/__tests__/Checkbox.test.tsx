import React from "react";
import { fireEvent, render, screen } from "@testing-library/react-native";

import Checkbox from "../Checkbox";

describe("Checkbox", () => {
  it("renders with default props", () => {
    render(<Checkbox>Default Checkbox</Checkbox>);
    expect(screen.getByText("Default Checkbox")).toBeTruthy();
    expect(screen.toJSON()).toMatchSnapshot();
  });

  it("renders with a secondary color variant", () => {
    render(<Checkbox color="secondary">Secondary</Checkbox>);
    expect(screen.getByText("Secondary")).toBeTruthy();
    expect(screen.toJSON()).toMatchSnapshot();
  });

  it("renders with a large size variant", () => {
    render(<Checkbox size="lg">Large Checkbox</Checkbox>);
    expect(screen.getByText("Large Checkbox")).toBeTruthy();
    expect(screen.toJSON()).toMatchSnapshot();
  });

  it("disables the checkbox", () => {
    render(<Checkbox disabled>Disabled Checkbox</Checkbox>);
    expect(screen.getByRole("checkbox").props.accessibilityState.disabled).toBe(
      true,
    );
  });

  it("handles value change correctly", () => {
    const onValueChange = jest.fn();
    render(<Checkbox onValueChange={onValueChange}>Toggle Me</Checkbox>);

    const checkbox = screen.getByRole("checkbox");
    fireEvent.press(checkbox);
    expect(onValueChange).toHaveBeenCalledWith(true);
  });

  it("matches snapshot for different states", () => {
    render(<Checkbox disabled>Disabled Checkbox</Checkbox>);
    expect(screen.toJSON()).toMatchSnapshot();
  });
});
