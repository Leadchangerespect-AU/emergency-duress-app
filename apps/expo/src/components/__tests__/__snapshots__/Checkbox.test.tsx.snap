// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox matches snapshot for different states 1`] = `
<View
  className="flex-row undefined"
>
  <View
    accessibilityRole="checkbox"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": true,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    className="mr-2 px-2.5 py-2.5 accent-gray-500"
    collapsable={false}
    focusable={true}
    id=":r5:"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    role="checkbox"
    style={
      [
        {
          "borderColor": "#657786",
          "borderRadius": 2,
          "borderWidth": 2,
          "height": 20,
          "width": 20,
        },
        undefined,
        undefined,
        {
          "backgroundColor": undefined,
          "borderColor": "#003951",
        },
        {
          "backgroundColor": "transparent",
          "borderColor": "#CCD6DD",
        },
        undefined,
      ]
    }
  />
  <Text
    className="text-sm text-gray-400"
  >
    Disabled Checkbox
  </Text>
</View>
`;

exports[`Checkbox renders with a large size variant 1`] = `
<View
  className="flex-row undefined"
>
  <View
    accessibilityRole="checkbox"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    className="mr-2 accent-primary px-4 py-4"
    collapsable={false}
    focusable={true}
    id=":r2:"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    role="checkbox"
    style={
      [
        {
          "borderColor": "#657786",
          "borderRadius": 2,
          "borderWidth": 2,
          "height": 20,
          "width": 20,
        },
        undefined,
        undefined,
        {
          "backgroundColor": undefined,
          "borderColor": "#003951",
        },
        undefined,
        undefined,
      ]
    }
  />
  <Text
    className="text-md"
  >
    Large Checkbox
  </Text>
</View>
`;

exports[`Checkbox renders with a secondary color variant 1`] = `
<View
  className="flex-row undefined"
>
  <View
    accessibilityRole="checkbox"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    className="mr-2 px-2.5 py-2.5"
    collapsable={false}
    focusable={true}
    id=":r1:"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    role="checkbox"
    style={
      [
        {
          "borderColor": "#657786",
          "borderRadius": 2,
          "borderWidth": 2,
          "height": 20,
          "width": 20,
        },
        undefined,
        undefined,
        {
          "backgroundColor": undefined,
          "borderColor": "#04c8c7",
        },
        undefined,
        undefined,
      ]
    }
  />
  <Text
    className="text-sm"
  >
    Secondary
  </Text>
</View>
`;

exports[`Checkbox renders with default props 1`] = `
<View
  className="flex-row undefined"
>
  <View
    accessibilityRole="checkbox"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    className="mr-2 accent-primary px-2.5 py-2.5"
    collapsable={false}
    focusable={true}
    id=":r0:"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    role="checkbox"
    style={
      [
        {
          "borderColor": "#657786",
          "borderRadius": 2,
          "borderWidth": 2,
          "height": 20,
          "width": 20,
        },
        undefined,
        undefined,
        {
          "backgroundColor": undefined,
          "borderColor": "#003951",
        },
        undefined,
        undefined,
      ]
    }
  />
  <Text
    className="text-sm"
  >
    Default Checkbox
  </Text>
</View>
`;
